Initial tests for tidy.py #9152

mskrzypkows opened the pull request:
I've created tests for spaces check and license. Tell me if it's good direction so I'll add more tests.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9835)

<!-- Reviewable:end -->

nox commented:
```
Checking files for tidiness...
./python/tidy_test/incorrect_license.rs:1: incorrect license
./python/tidy_test/wrong_space.rs:5: trailing whitespace
./python/tidy_test/wrong_space.rs:7: no newline at EOF
./python/tidy_test/wrong_space.rs:7: tab on line
./python/tidy_test/wrong_space.rs:7: CR on line
./python/tidy_test/wrong_space.rs:8: no newline at EOF
```

nox commented:
Seems like the test fixtures shouldn't be linted when linting the actual repository and not running Tidy tests.

Wafflespeanut commented:
Ah! That could simply be achieved by adding the "tidy_tests" to `ignored_dirs` :)

@mskrzypkows I had a brief overview at your PR, and I think you're in the right direction. The tests could be enhanced, but let's focus on the testsuite for now (I'll review it ASAP).

Wafflespeanut commented on python/tidy_test/test_tidy.py:
```diff
@@ -0,0 +1,73 @@
+    print("USAGE: python {0}".format(sys.argv[0]))
+
+
+if __name__ == '__main__':
```
This should be wrapped inside a function, so that `mach` can call it whenever we invoke the tidy testing command.

Wafflespeanut commented on python/tidy.py:
```diff
@@ -555,6 +555,9 @@ def get_wpt_files(only_changed_files, progress):
 def check_wpt_lint_errors(files):
     wpt_working_dir = os.path.abspath(os.path.join(".", "tests", "wpt", "web-platform-tests"))
+    # check if run by test-tidy script form main directory or from test_tidy.py script, from python/tidy_test
+    if not os.path.isdir(wpt_working_dir):
```
I presume this is going to be called by `mach` anyway (probably a new command? /cc @jdm), and so we don't want this.

mskrzypkows commented:
OK, so I'll create more test suite.

Wafflespeanut commented:
@jdm Shall we have a new command argument (say, `test-tidy --self`), so that this test would be called by `mach` itself, and we don't have to do all this)?

mskrzypkows commented:
Should I test check_flake8 function? Inside check_flake8 there is stdout redirect which conflicts with my stdout redirect in CheckRustTidiness class. Maybe there is no need to check python files because they're checked by external library. How do you think?

mskrzypkows commented:
Added more tests, it's not prepared for `mach` command yet. Waiting for decision.

Wafflespeanut commented:
Yep, I think we can ignore the flake8 tests for now. And, let's go for `./mach test-tidy --self`. I'll look into your recent changes soon :)

Wafflespeanut commented:
This looks good so far (/cc @jdm)

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 3 of 4 files at r1, 6 of 7 files at r2.
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[python/tidy.py, line 559 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9835#-KBqhgLyqbxWWBatkZFP-r1-559:-KCUQcVP8NL9t0NZKpYC:-1954599262) ([raw file](https://github.com/servo/servo/blob/e334f13658c9dc926b107c3fa0c02d021e0115e7/python/tidy.py#L559)):*
We should remove this (`mach` will handle this for us - see below)

---

*[python/tidy_test/tidy_test.py, line 113 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9835#-KCqQ9UXn-N0rrk43bN8:-KCqQ9UXn-N0rrk43bN9:-1750238770) ([raw file](https://github.com/servo/servo/blob/482701b66848d5b9ef3968aa1969d13ea962f51b/python/tidy_test/tidy_test.py#L113)):*
This hasn't been addressed yet. We should [add a new argument](https://github.com/servo/servo/blob/455ed0ae6d4c82fd38762a92f0dd6697bd3b2b50/python/servo/testing_commands.py#L293) to `mach test-tidy`(say, `--self`), and invoke the testing function from there (so that we don't have to worry about some redundant code - tidy, bootstrapping, etc.).

---

*[python/tidy_test/wrong_space.rs, line 7 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9835#-KCqPkX6tm8IUgkVWbP_:-KCqPkX7rLnnrv0UGLLg:1918481386) ([raw file](https://github.com/servo/servo/blob/482701b66848d5b9ef3968aa1969d13ea962f51b/python/tidy_test/wrong_space.rs#L7)):*
Nit: newline at EOF?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9835#-:-KC-kl8YdrbDqULAe83C:-54652602)*
<!-- Sent from Reviewable.io -->

mskrzypkows commented:
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[python/tidy_test/wrong_space.rs, line 7 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9835#-KCqPkX6tm8IUgkVWbP_:-KCtEUeCzG1lSEeiUQ_X:-881393626) ([raw file](https://github.com/servo/servo/blob/482701b66848d5b9ef3968aa1969d13ea962f51b/python/tidy_test/wrong_space.rs#L7)):*
It's intentional to check if tests detect that there is no new line at EOF.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9835)*
<!-- Sent from Reviewable.io -->

Wafflespeanut commented:
Reviewed 1 of 7 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9835)*
<!-- Sent from Reviewable.io -->

mskrzypkows commented:
I've integrated tests with mach command, they are run by `./mach test-tidy --self-test`.

Wafflespeanut commented:
-S-awaiting-review +S-needs-code-changes 

With this setup on hand, let's add `./mach test-tidy --self-test` to `.travis.yml` before actually invoking the tidy test (so that we can see how this turns out in the CI builder)

---

Reviewed 15 of 16 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[python/tidy.py, line 41 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9835#-KD2ux9JDzEK3987BN4z:-KD2ux9JDzEK3987BN5-:-812636473) ([raw file](https://github.com/servo/servo/blob/c81da92893eed6626121f21caa3eba79bc5778d0/python/tidy.py#L41)):*
We won't be needing this, since the parent directory has already been added to `ignored_dirs`

---

*[python/tidy.py, line 611 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9835#-KD2vLWl_WfrL-76H9nF:-KD2vLWl_WfrL-76H9nG:2040873718) ([raw file](https://github.com/servo/servo/blob/c81da92893eed6626121f21caa3eba79bc5778d0/python/tidy.py#L611)):*
Instead of modifying tidy for the sake of the testsuite and checking the output, shall we just import the stuff from tidy and check the yielded values within the testsuite itself?

By doing that, we won't have to rely on `scan`, and we also won't have to capture the output.

---

*[python/servo/testing_commands.py, line 34 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9835#-KD2tgpSVTYtZEcn7c9S:-KD2tgpTISgNAKCzIkDT:2123651720) ([raw file](https://github.com/servo/servo/blob/c81da92893eed6626121f21caa3eba79bc5778d0/python/servo/testing_commands.py#L34)):*
Nit: Shall we go for the [built-in function](https://docs.python.org/2/library/site.html#site.addsitedir) instead? It looks somewhat cleaner.

---

*[python/servo/testing_commands.py, line 104 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9835#-KD3-Bh5oCOCNg2gY5PZ:-KD3-Bh5oCOCNg2gY5P_:755317851) ([raw file](https://github.com/servo/servo/blob/c81da92893eed6626121f21caa3eba79bc5778d0/python/servo/testing_commands.py#L104)):*
Nit: Please change this to "Run unit tests for tidy" (it's not a command, it's our pet) :smile:

---

*[python/servo/testing_commands.py, line 304 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9835#-KD3-TrZn8bCgUPPjtLr:-KD3-Tr_7I2cVxA3LnXd:-469765787) ([raw file](https://github.com/servo/servo/blob/c81da92893eed6626121f21caa3eba79bc5778d0/python/servo/testing_commands.py#L304)):*
ditto

---

*[python/tidy_self_test/tidy_self_test.py, line 87 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9835#-KD3-Z-OJYobTFnd-qCV:-KD3-Z-OJYobTFnd-qCW:1169179175) ([raw file](https://github.com/servo/servo/blob/c81da92893eed6626121f21caa3eba79bc5778d0/python/tidy_self_test/tidy_self_test.py#L87)):*
ehm, any plans to finish this off?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9835#-:-KD22ZKjaGpw_9J7N9XC:1252514164)*
<!-- Sent from Reviewable.io -->

mskrzypkows commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[python/tidy.py, line 41 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9835#-KD2ux9JDzEK3987BN4z:-KE5Tkm_QEY7lyWN4g01:-1685941120) ([raw file](https://github.com/servo/servo/blob/c81da92893eed6626121f21caa3eba79bc5778d0/python/tidy.py#L41)):*
It's added because when I run test for 'tidy_self_test' directory, tidy_self_test.py was not ignored. Probably when I change not to use scan function, it could be removed.

---

*[python/tidy_self_test/tidy_self_test.py, line 87 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9835#-KD3-Z-OJYobTFnd-qCV:-KE5VYubqNcCQRreiVia:1521120647) ([raw file](https://github.com/servo/servo/blob/c81da92893eed6626121f21caa3eba79bc5778d0/python/tidy_self_test/tidy_self_test.py#L87)):*
Yes, do you know situation when "extern crate declaration" error appears?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9835)*
<!-- Sent from Reviewable.io -->

mskrzypkows commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[python/tidy.py, line 559 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9835#-KBqhgLyqbxWWBatkZFP-r1-559:-KE6KGtlZuAVCcP7aLcj:-238128543) ([raw file](https://github.com/servo/servo/blob/e334f13658c9dc926b107c3fa0c02d021e0115e7/python/tidy.py#L559)):*
Should I remove whole function?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9835)*
<!-- Sent from Reviewable.io -->

Wafflespeanut commented:
-S-awaiting-review +S-needs-code-changes 

We're looking good so far. Still a few more :)

btw, my last comment on modifying `.travis.yml` hasn't been addressed yet.

---

Reviewed 1 of 16 files at r3, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[python/tidy_self_test/tidy_self_test.py, line 87 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9835#-KD3-Z-OJYobTFnd-qCV:-KEVk0GkYq4eGzepkPLh:-2008313870) ([raw file](https://github.com/servo/servo/blob/c81da92893eed6626121f21caa3eba79bc5778d0/python/tidy_self_test/tidy_self_test.py#L87)):*
It occurs when the `mod`, `use` and `extern crate` declarations aren't sorted. Try running tidy after swapping a few declaration lines in a Rust source file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9835#-:-KEUf1Si9imBpQsdLbP8:362393807)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #10074) made this pull request unmergeable. Please resolve the merge conflicts.

mskrzypkows commented:
Addressed your comments. There is still lack of tests for 
```
check_spec
check_wpt_lint_errors
check_lock
```
functions. Can that be made in separate commit? I want to have merged current state of tests :-)

---

Review status: 5 of 11 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9835#-:-KEf9DTS_K7_b2iUOudg:-741715021)*
<!-- Sent from Reviewable.io -->

Wafflespeanut commented:
Of course! Just push another commit to the same branch (We can merge a series of commits in the same PR). Also, please make sure that a future commit doesn't revert the changes made in a previous commit.

-S-awaiting-review -S-needs-rebase +S-needs-code-changes 

---

Reviewed 2 of 16 files at r3, 5 of 6 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9835#-:-KEjQTJiNDzQDCw5gLju:1991245580)*
<!-- Sent from Reviewable.io -->

mskrzypkows commented:
I want to merge this single commit to have it updated when tidy.py is changed. Rest I'll do next week, it could be another PR. Is it possible to merge this commit in current state?

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9835#-:-KEjg2ZKlNamXsDS6RDV:1915113952)*
<!-- Sent from Reviewable.io -->

Ms2ger commented on python/servo/testing_commands.py:
```diff
@@ -31,6 +31,10 @@
 import tidy
 
+import site
+site.addsitedir(os.path.abspath(os.path.join(os.path.dirname(__file__), '../tidy_self_test/')))
```
Is this really the best approach? @frewsxcv?

frewsxcv replied:
In this particular case, I would recommend just turning `python/tidy_self_test/` into a Python module via `touch python/tidy_self_test/__init__.py`. That *should* allow you to do the `import tidy_self_test` here without any `sys.path` overrides.

mskrzypkows replied:
Done

Commit:
tests for tidy.py #9152

jdm commented:
I am definitely in favour of merging some tests rather than waiting for all tests.

Wafflespeanut commented:
Sorry, I misunderstood. Thanks for doing this :)

@bors-servo r+

---

Reviewed 1 of 16 files at r3, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9835#-:-KEl3j02qQSN7_k6Ahtx:833054474)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit b8c9a28 has been approved by `Wafflespeanut`

<!-- @bors-servo r=Wafflespeanut b8c9a288e9cd6dea04082f997fdd0669ddea9b4e -->

bors-servo commented:
:hourglass: Testing commit b8c9a28 with merge 72eae39...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/168)

frewsxcv commented:
@bors-servo retry https://github.com/servo/servo/issues/8931

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/148), [arm32](http://build.servo.org/builders/arm32/builds/161), [arm64](http://build.servo.org/builders/arm64/builds/154), [linux-dev](http://build.servo.org/builders/linux-dev/builds/156), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/161), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/154), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/158) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/168)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/148), [arm32](http://build.servo.org/builders/arm32/builds/161), [arm64](http://build.servo.org/builders/arm64/builds/154), [linux-dev](http://build.servo.org/builders/linux-dev/builds/156), [linux-rel](http://build.servo.org/builders/linux-rel/builds/170), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/161), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/154), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/158)
